core/state_connector.go

```diff
@@ -14,219 +15,204 @@ import (
+		h, err := c.attestationResult(attestors[i], instructions)
+		if err != nil {
+			av.abstainedAttestors = append(av.abstainedAttestors, attestors[i])
+		}
+		hashFrequencies[h] = append(hashFrequencies[h], attestors[i])
+	}
+	// Find the plurality
+	var pluralityNum int
+	var pluralityKey string
+	for key, val := range hashFrequencies {
+		if len(val) > pluralityNum {
+			pluralityNum = len(val)
+			pluralityKey = key
+		}
 	}
-	return false
+	if pluralityNum > len(attestors)/2 {
+		av.reachedMajority = true
+		av.majorityDecision = pluralityKey
+		av.majorityAttestors = hashFrequencies[pluralityKey]
+	}
+	for key, val := range hashFrequencies {
+		if key != pluralityKey {
+			av.divergentAttestors = append(av.divergentAttestors, val...)
+		}
+	}
+	return av
 }
 
-func GetStateConnectorContract(chainID *big.Int, blockTime *big.Int) common.Address {
-	switch {
-	case GetStateConnectorActivated(chainID, blockTime) && chainID.Cmp(songbirdChainID) == 0:
-		return common.HexToAddress("0x6b5DEa84F71052c1302b5fe652e17FD442D126a9")
-	default:
-		return common.HexToAddress("0x1000000000000000000000000000000000000001")
+// defaultAttestors returns list of FTSO price providers which represents the default attestors.
+func (c *stateConnector) defaultAttestors(chainID *big.Int, timestamp *big.Int) ([]common.Address, error) {
+	bc := c.caller.BlockContext()
+	// Get VoterWhitelister contract
+	voterWhitelisterContractBytes, _, err := c.caller.Call(
+		vm.AccountRef(c.msg.From()),
+		common.HexToAddress(GetPrioritisedFTSOContract(timestamp)),
+		voterWhitelisterSelector(chainID, timestamp),
+		GetKeeperGasMultiplier(bc.BlockNumber)*bc.GasLimit,
+		big.NewInt(0))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get VoterWhitelister contract: %w", err)
+	}
+
+	// Get FTSO price providers
+	voterWhitelisterContract := common.BytesToAddress(voterWhitelisterContractBytes)
+	priceProvidersBytes, _, err := c.caller.Call(
+		vm.AccountRef(c.msg.From()),
+		voterWhitelisterContract,
+		ftsoWhitelistedPriceProvidersSelector(chainID, timestamp),
+		GetKeeperGasMultiplier(bc.BlockNumber)*bc.GasLimit,
+		big.NewInt(0))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get FTSO price providers: %w", err)
 	}
+
+	attestorsNum := len(priceProvidersBytes) / common.HashLength
+	var attestors []common.Address
+	for i := 0; i < attestorsNum; i++ {
+		attestors = append(attestors, common.BytesToAddress(priceProvidersBytes[i*common.HashLength:(i+1)*common.HashLength]))
+	}
+	return attestors, nil
 }
 
-func GetStateConnectorCoinbaseSignalAddr(chainID *big.Int, blockTime *big.Int) common.Address {
-	switch {
-	default:
-		return common.HexToAddress("0x000000000000000000000000000000000000dEaD")
+func (c *stateConnector) attestationResult(attestor common.Address, instructions []byte) (string, error) {
+	merkleRootHash, _, err := c.caller.Call(vm.AccountRef(attestor), c.to(), instructions, 20000, big.NewInt(0))
+	return hex.EncodeToString(merkleRootHash), err
```

Comment:
I'd argue that we can drop the `merkle` here without losing any meaning. It's a root hash, doesn't really matter if it's the root hash for a merkle tree or even a patricia merkle one IMO